### DISCUSSION TITLE
postfixadmin stopped working after upgrade - strange php errors. Help please!

### DISCUSSION THREAD

**Discussion by jrcdec** [postfixadmin/postfixadmin#797] [General]
I have a fairly mature Postfix server which has been running postfixadmin since 2010. After a recent update, postifxadmin has stopped working as shown by the two attached screenshots.
The server runs normally for existing mail accounts, but I am unable to add new ones or manage the existing ones. I don't know if this happened with the latest update or the one before that because our user list had been stable for the last 3 years and I only found out when suddenly trying to add more users.
The first error about not using https: is false - the screenshot is the result of accessing the setup screen over https,
I have tried addressing the second error about the changed password format by entering the new format password hash in config.local.php but it had no effect - the same error comes up in the postfixadmin setup screen.
Any suggestions for what to try next will be more than welcome - I suspect the slew of php errors in the postfixadmin setup screen might look familiar to someone and suggest what the culprit new setting or permission might be.
The system is running Ubuntu 22.04 jammy on kernel 5.15.0-91,php 8.1.2, postfix, dovecot, and mysql, with spamassassin as the spam filter and all packages on the latest version in the ubuntu stable release. postfixadmin should therefore be  3.3.10-2 but I can't tell for sure as I can't get to the login screen to see it
![Postfixadmin1](https://github.com/postfixadmin/postfixadmin/assets/156660371/0f66d38a-16be-4ba3-ae36-c8a8a318bb43)
![Postfixadmin2](https://github.com/postfixadmin/postfixadmin/assets/156660371/3914aa03-8e54-4106-af12-12e3c93e02d1)

Many thanks in advance

James
(reposting from sourceforge earlier today since I just learned postfixadmin questions now live here)

**Comment by DavidGoodwin** [comment]
I think you're using a pretty old variant of PostfixAdmin, and your PHP runtime has changed to be somewhat newer.

You can find out your PostfixAdmin version by looking in config.inc.php for something like: `$CONF['version'] = ....`. Alternatively, if you've installed through your distribution's package tool - `dpkg -l postfixadmin` 


You can either try and update the postfixadmin code, or try and change the PHP config. Changing PHP's config is probably easiest. If you are using PHP 8.1 with an old version of PostfixAdmin you'll probably encounter other problems.


 1. Update PostfixAdmin outside of the distribution's package by downloading the latest 3.3 tar file from https://github.com/postfixadmin/postfixadmin/releases .... you'll probably need to update your apache config file as well as accessing 'setup.php' in your browser to apply db updates. You're likely to get some confusion over which config file is in use / which source code is being run, so this is probably a more difficult fix.
 
 2. Update PHP's settings to cope with your legacy Postfixadmin install - create a file within the web root named something like 'phpinfo.php' containing '<?php phpinfo();' and access it through your browser; it will confirm the version of PHP you think you're using and also show the setting for the short_open_tags parameter. Then, edit /etc/php/<version>/apache2/php.ini and add a line on the bottom for 'short_open_tags = On' ( I think the version of Postfixadmin you're trying to use is too old, and hasn't had various short open tag fixes ). This might be the easier fix?

 3. Downgrade PHP to PHP 7.4 ... this might fix it? Depends on the server config though.

**Reply by jrcdec** [reply]
Thanks.
dpkg -l postfixadmin confirms I am using version 3.3.10, which is fairly up to date (and is the correct current version for Ubuntu 22.04 server), so it seems we are not dealing with an ancient version
php --version shows the current version is 8.1.2 which is also very current and the right version for Ubuntu 22.04
I have changed short_open_tags from Off to On in php.ini but I still get the exact same output on postfixadmin/setup.php as in the original message. I suspect something else might be the problem.
Many thanks again - look forward to your next idea to try out...

**Reply by DavidGoodwin** [reply]
Are you sure you don't have multiple installations of postfixadmin ? (as in you've got the .deb installed which puts the files in /some/where and you've got a manual install which has them in /some/blah .... and your webserver is reading from /some/blah?)

**Reply by jrcdec** [reply]
These are all the postfixadmin related entries I could find using "find":
find / -name postfixadmin
[sudo] password for jrc: 
/usr/share/doc/postfixadmin
/usr/share/postfixadmin
/var/cache/postfixadmin
/var/lib/mysql/postfixadmin
/var/lib/apache2/conf/enabled_by_maint/postfixadmin
/etc/apache2/postfixadmin
/etc/postfixadmin

All upgrades since 2010 have been done withe the regular Ubuntu package manager.
They look like plausible results of a package installation - do any look out of line to you?

Thanks!

**Reply by jrcdec** [reply]
(the original 2010 installation was also from the Ubuntu package as opposed to manual)

**Reply by DavidGoodwin** [reply]
Checking what the webserver's config is might be more reliable than hoping the directory name matches.

Are you able to revert to using PHP 7.4? Does that make a difference?

**Reply by jrcdec** [reply]
going back to php 7.4 (the version in ubuntu 20.04) It would be messy, because the rest of the installation is all on package versions consistent with the current installed release of Ubuntu Server, which is 22.04 so I would need to downgrade and hope nothing else breaks. 
I am checking the postfixadmin directories one by one and this is what I am finding so far:
/usr/share/doc/postfixadmin - the files are mostly dated 9 Aug 2021. Some directories are dated 12 Jan 2024 but the files in them are also 9 Aug 2021
/usr/share/postfixadmin  - ditto - all files 9 Aug 2021, including those in the scripts and model subdirectories etc.
/var/cache/postfixadmin - this only contains an empty subdirectory called templates_c, so no old version files here
/var/lib/mysql/postfixadmin - this is empty so no old version files here
/var/lib/apache2/conf/enabled_by_maint/postfixadmin  - this is an empty file (0 bytes), permissions -rw-r--r-- 1 root root , dated 25 april 2016
/etc/apache2/postfixadmin - this is a short (42 byte) file owned rw by www-data only and containing what looks like a password hash. It is dated 26 April 2016 so could it potentially be a now-incompatible hash format?
and finally /etc/postfixadmin which contains the important config.inc.php, config.local.php, and dbconfig.inc.php  The first two have recent dates and the dbconfig one is dated 2016.
The contents are : 
'<?php
##
## database access settings in php format
## automatically generated from /etc/dbconfig-common/postfixadmin.conf
## by /usr/sbin/dbconfig-generate-include
##
## by default this file is managed via ucf, so you shouldn't have to
## worry about manual changes being silently discarded.  *however*,
## you'll probably also want to edit the configuration file mentioned
## above too.
##
$dbuser='postfix';
$dbpass='(removed by me before posting)';
$basepath='';
$dbname='postfix';
$dbserver='localhost';
$dbport='';
$dbtype='mysqli';
#$dbtype='mysql';

'Does it need some sort of ?> ending to match the <?php opening ?

I tried one more thing and got a possibly informative error:
sudo postfixadmin-cli help

ERROR: The PostfixAdmin database layout is outdated (you have r1841, but r1844 is expected).
Please run setup.php to upgrade the database.

Does any of this help?

**Reply by jrcdec** [reply]
More possible clues. If I try to long in to https://servername/postfixadmin/login.php, I get the following output on the web page that pops up.
This is odd because it complains about there being no configuration in config.local.php yet the original configuration is still there. 
How does one update teh database format from the old to the new format?

getAll(); $smarty = PFASmarty::getInstance(); if ($CONF['configured'] !== true) { print "Installation not yet configured; please edit config.inc.php or write your settings to config.local.php"; exit; } check_db_version(); # check if the database layout is up to date (and error out if not) if ($_SERVER['REQUEST_METHOD'] == "POST") { if (!isset($_SESSION['PFA_token'])) { die("Invalid token (session timeout; refresh the page and try again?)"); } if (safepost('token') != $_SESSION['PFA_token']) { die('Invalid token! (CSRF check failed)'); } $lang = safepost('lang'); $fUsername = trim(safepost('fUsername')); $fPassword = safepost('fPassword'); if ($lang != check_language(false)) { # only set cookie if language selection was changed setcookie('lang', $lang, time() + 60*60*24*30); # language cookie, lifetime 30 days # (language preference cookie is processed even if username and/or password are invalid) } $h = new AdminHandler(); $login = new Login('admin'); if ($login->login($fUsername, $fPassword)) { init_session($fUsername, true); # they've logged in, so see if they are a domain admin, as well. if (!$h->init($fUsername)) { flash_error($PALANG['pLogin_failed']); } if (!$h->view()) { flash_error($PALANG['pLogin_failed']); } $adminproperties = $h->result(); if ($adminproperties['superadmin'] == 1) { $_SESSION['sessid']['roles'][] = 'global-admin'; } header("Location: main.php"); exit(0); } else { # $h->login failed error_log("PostfixAdmin admin login failed (username: $fUsername, ip_address: {$_SERVER['REMOTE_ADDR']})"); flash_error($PALANG['pLogin_failed']); } } else { session_unset(); session_destroy(); session_start(); } $_SESSION['PFA_token'] = md5(uniqid("pfa" . rand(), true)); $smarty->assign('language_selector', language_selector(), false); $smarty->assign('smarty_template', 'login'); $smarty->assign('logintype', 'admin'); $smarty->assign('forgotten_password_reset', Config::bool('forgotten_admin_password_reset')); $smarty->display('index.tpl'); /* vim: set expandtab softtabstop=4 tabstop=4 shiftwidth=4: */

**Reply by DavidGoodwin** [reply]
is the problem just that your webserver isn't configured to handle PHP ? Do any /etc/apache2/mods-enabled/php*.load files exist?
(assuming you're using mod_php and not e.g. fpm)

( hint: a2enmod php8.1; service apache2 restart )

**Reply by jrcdec** [reply]
Many thanks - that did it!
I had originally used a2enmod to enable php, but for some reason the system had updated php from 7.4 to 8.1 but only php7.4 was "a2enmod-ed" after the update - it failed to reflect the new php in the enabled modules folder. So I did a2dismod php7.4, a2enmod php8.1 and the setup screen magically works now.

**Reply by jrcdec** [reply]
![Postfixadmin3](https://github.com/postfixadmin/postfixadmin/assets/156660371/2206f0e0-675a-4b2b-97be-9186be05c031)

**Reply by jrcdec** [reply]
Since I am only using mysql I presume I can safely ignore the postgresql and SQLite warnings?

**Reply by DavidGoodwin** [reply]
phew, sometimes the obvious solutions aren't so obvious :)

Yes you can ignore the postgresql/sqlite things - they're just notices really.

**Reply by jrcdec** [reply]
Many thanks again! I strongly suspected it had to be a php problem from the spurious php leaking onto the webpage, but i did not think of something as simple as a2enmod having been overlooked by the package update from 7.4 to 8.1. 
It took you to spot it.
Separately, I have seen some chat on the issue of all the noise coming out of some deprecated php functions complaining when fed null arguments. Are you planning to replace those at some point? I suspect our current workaround of dialing down the log noise level in php.ini will stop working at some point in the future...

**Reply by DavidGoodwin** [reply]
I'm not aware of any - if you can give me the error message(s) and what version of PostfixAdmin you're using, I can fix them easily enough.

**Comment by mmbret** [comment]
Is this thread still active? I have a couple postfixadmin questions similar to this thread.
Thanks Bret
